src/ArmTemplates/Common/Templates/TemplateBuilder.cs

```diff
@@ -0,0 +1,108 @@
+﻿// --------------------------------------------------------------------------
+//  <copyright file="TemplateBuilder.cs" company="Microsoft">
+//      Copyright (c) Microsoft Corporation. All rights reserved.
+//  </copyright>
+// --------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Constants;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Templates.Abstractions;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Extractor.Models;
+
+namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Templates
+{
+    static class TemplateBuilder
```

Comment:
Again, not a fan of making everything static. We can't do proper unit tests with static stuff. I would either use inheritance or DI.

Reply:
I just decided to make it as a fluent API for creating template, so that you can do something like this:
```
var template = 
TemplateBuilder.CreateEmpty()
                          .AddParameter1()
                          .AddParameter2()
                          .Add....
                          ....()
```

Because I see that in every extractor there is some call going to base class, but then there are some additional modifications. I don't see how this functionality is mocked - it is just a creation of object but with additional properties set.

So basic api for template generation will be available only in inherited classes, however there will be a possibility to use this fluent API for creating templates and easily constructing one you want without copy&paste code

Reply:
I like the fluent API. I just don't like that it's static. Why not just pass an ITemplateBuilder via DI? That way you can mock it in tests if you want to. If you don't want to mock it then don't. There's no downside.

> I don't see how this functionality is mocked - it is just a creation of object but with additional properties set.
It doesn't really matter what the class does. Every class should have its own independent unit tests. When you're unit testing a class, you shouldn't be running the code of another class. You should just mock it so that your tests for class X don't break when someone changes class Y.

Reply:
It was painful, but I did it :)
Please, view how it is working now - `ITemplateBuilder`